marshmallow/fields.py

```diff
@@ -531,12 +531,26 @@ def _serialize(self, value, attr, obj):
         return [self.container._serialize(value, attr, obj)]
 
     def _deserialize(self, value, attr, data):
-        if utils.is_collection(value):
-            # Convert all instances in typed list to container type
-            return [self.container.deserialize(each) for each in value]
-        else:
+        if not utils.is_collection(value):
             self.fail('invalid')
 
+        result = []
+        errors = {}
+        for idx, each in enumerate(value):
+            try:
+                result.append(self.container.deserialize(each))
+            except ValidationError as e:
+                result.append(e.data)
+                item_errors = e.messages
+                if utils.is_collection(item_errors) and len(item_errors) == 1:
+                    item_errors = item_errors[0]
+                errors.update({str(idx): item_errors})
```

Comment:
I don't think there is a need to cast `idx` to a string. When passing `many=True` to a `Schema`, errors are indexed by integer values, so best to keep that consistent here.